### PR TITLE
V4.0 instead of master

### PR DESCRIPTION
**Pull request by lbarman**
(not ready for merge)

### CODE REVIEW

**Comment by hmil** [comment]
Why rename dto -> messages ? [DTO](https://en.wikipedia.org/wiki/Data_transfer_object) actually describes a design pattern that is often used in the industry.

**Comment by lbarman** [comment]
I prefer `messages`. I find it easier to see the `server` as a stateful thing that reacts to `messages`, like a protocol.

File: packages/vaultage-ui-webcli/src/commands/Edit.ts [hunk 1]
```diff
@@ -56,7 +56,14 @@ export class EditCommand implements ICommand {
             this.shell.echoHTML(VaultEntryFormatter.formatSingle(entry2))
             this.shell.echo("Updated entry #"+id)
 
-            await new Promise((_resolve, reject) => this.vault.save(reject));
+            await new Promise((resolve, reject) => this.vault.save(function(err) {
+                if(err == null){
+                    resolve()
+                } else {
+                    reject('<span class="error">'+err.toString()+'</span>')
```
Comment (hmil) [inline packages/vaultage-ui-webcli/src/commands/Edit.ts]: `echoErro` already wraps the text in a `<span class="error">` so we don't have to concatenate strings all over the business logic.

Comment (hmil) [reply packages/vaultage-ui-webcli/src/commands/Edit.ts]: Was there an issue with the compact form?

File: packages/vaultage/src/storage/JSONDatabase.ts [hunk 2]
```diff
@@ -30,12 +30,17 @@ export class JSONDatabase implements IDatabase {
     }
 
     public async save(update: IDatabaseSaveParameters): Promise<string> {
+        let password = this.password;
+        if (update.new_password !== undefined) {
```
Comment (hmil) [inline packages/vaultage/src/storage/JSONDatabase.ts]: Could be more fault-tolerant with `!= null` instead (covers both `null` and `undefined`)

**Comment by lbarman** [comment]
Fiesta!

**Pull request merged**